File: internal/diff/diff.go
// Package diff implements the noise-aware response diff and write-set diff.
package diff

import (
	"encoding/base64"
	"encoding/json"
	"sort"
	"strconv"
	"strings"

	"github.com/shopspring/decimal"

	"github.com/pajamasi726/mocking-box/internal/writeset"
)

// Verdicts.
const (
	Match        = "MATCH"
	ResponseDiff = "RESPONSE_DIFF"
	WritesetDiff = "WRITESET_DIFF"
	BothDiff     = "BOTH_DIFF"
	// Noise: every difference matched the baseline (self-check) fingerprints —
	// a capture artifact, not a behavior change of the new stack.
	Noise = "NOISE"
	Error = "ERROR"
)

type Difference struct {
	Kind string `json:"kind"` // "response" | "writeset"
	Path string `json:"path"`
	Old  any    `json:"old"`
	New  any    `json:"new"`
}

const absent = "<absent>"

// ---------------------------------------------------------------------------
// value normalization (binlog & JSON values -> comparable primitives)
// ---------------------------------------------------------------------------

func NormValue(v any) any {
	switch x := v.(type) {
	case decimal.Decimal:
		// scale-insensitive: 50000 == 50000.0
		if x.Equal(x.Truncate(0)) {
			return x.IntPart()
		}
		f, _ := x.Float64()
		return f
	case []byte:
		if isPrintable(x) {
			return string(x)
		}
		return base64.StdEncoding.EncodeToString(x)
	case float64:
		// JSON numbers arrive as float64; canonicalize integral values
		if x == float64(int64(x)) {
			return int64(x)
		}
		return x
	case int:
		return int64(x)
	case int8:
		return int64(x)
	case int16:
		return int64(x)
	case int32:
		return int64(x)
	case uint:
		return int64(x)
	case uint8:
		return int64(x)
	case uint16:
		return int64(x)
	case uint32:
		return int64(x)
	case uint64:
		return int64(x)
	case float32:
		return NormValue(float64(x))
	case map[string]any:
		out := make(map[string]any, len(x))
		for k, val := range x {
			out[k] = NormValue(val)
		}
		return out
	case []any:
		out := make([]any, len(x))
		for i, val := range x {
			out[i] = NormValue(val)
		}
		return out
	default:
		return v
	}
}

func isPrintable(b []byte) bool {
	for _, c := range b {
		if c < 0x09 {
			return false
		}
	}
	return true
}

func canonical(v any) string {
	b, err := json.Marshal(NormValue(v)) // encoding/json sorts map keys
	if err != nil {
		return "!" + strconv.Quote(err.Error())
	}
	return string(b)
}

func equal(a, b any) bool { return canonical(a) == canonical(b) }

// ---------------------------------------------------------------------------
// noise path matching for JSON bodies
//   pattern segments: literal | * (exactly one) | ** (zero or more)
// ---------------------------------------------------------------------------

func PathMatches(pattern string, path []string) bool {
	return match(strings.Split(pattern, "."), path)
}

func match(pat, path []string) bool {
	if len(pat) == 0 {
		return len(path) == 0
	}
	head, rest := pat[0], pat[1:]
	if head == "**" {
		for i := 0; i <= len(path); i++ {
			if match(rest, path[i:]) {
				return true
			}
		}
		return false
	}
	if len(path) == 0 {
		return false
	}
	if head == "*" || head == path[0] {
		return match(rest, path[1:])
	}
	return false
}

// SortRule sorts arrays whose path matches Path before comparison, so mapper-
// dependent ordering (Jackson vs MyBatis result order, …) doesn't read as a diff.
// By is an element key to sort by, or "$canonical" to sort by the element's
// canonical JSON.
type SortRule struct {
	Path string `json:"path" yaml:"path"`
	By   string `json:"by" yaml:"by"`
}

// SortArrays returns a copy of v with matching arrays canonically sorted.
func SortArrays(v any, rules []SortRule) any { return sortArrays(v, rules, nil) }

func sortArrays(v any, rules []SortRule, path []string) any {
	switch x := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(x))
		for k, val := range x {
			out[k] = sortArrays(val, rules, append(append([]string{}, path...), k))
		}
		return out
	case []any:
		items := make([]any, len(x))
		for i, val := range x {
			items[i] = sortArrays(val, rules, append(append([]string{}, path...), strconv.Itoa(i)))
		}
		for _, rule := range rules {
			if !PathMatches(rule.Path, path) {
				continue
			}
			sortKey := func(e any) string {
				if rule.By != "" && rule.By != "$canonical" {
					if m, ok := e.(map[string]any); ok {
						return canonical(m[rule.By])
					}
				}
				return canonical(e)
			}
			sort.SliceStable(items, func(i, j int) bool { return sortKey(items[i]) < sortKey(items[j]) })
			break // first matching rule wins
		}
		return items
	default:
		return v
	}
}

// StripNoise returns a copy of v with noise-matching keys removed.
func StripNoise(v any, patterns []string) any { return stripNoise(v, patterns, nil) }

func stripNoise(v any, patterns []string, path []string) any {
	switch x := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(x))
	keys:
		for k, val := range x {
			child := append(append([]string{}, path...), k)
			for _, p := range patterns {
				if PathMatches(p, child) {
					continue keys
				}
			}
			out[k] = stripNoise(val, patterns, child)
		}
		return out
	case []any:
		out := make([]any, len(x))
		for i, val := range x {
			out[i] = stripNoise(val, patterns, append(append([]string{}, path...), strconv.Itoa(i)))
		}
		return out
	default:
		return v
	}
}

// ---------------------------------------------------------------------------
// response diff
// ---------------------------------------------------------------------------

func tryJSON(body string) (any, bool) {
	var v any
	if err := json.Unmarshal([]byte(body), &v); err != nil {
		return nil, false
	}
	switch v.(type) {
	case map[string]any, []any:
		return v, true
	}
	return v, false
}

func diffJSON(old, new any, path []string, out *[]Difference) {
	joined := func() string {
		if len(path) == 0 {
			return "$"
		}
		return strings.Join(path, ".")
	}
	oldMap, oldIsMap := old.(map[string]any)
	newMap, newIsMap := new.(map[string]any)
	if oldIsMap && newIsMap {
		keys := map[string]bool{}
		for k := range oldMap {
			keys[k] = true
		}
		for k := range newMap {
			keys[k] = true
		}
		sorted := make([]string, 0, len(keys))
		for k := range keys {
			sorted = append(sorted, k)
		}
		sort.Strings(sorted)
		for _, k := range sorted {
			child := append(append([]string{}, path...), k)
			ov, oOK := oldMap[k]
			nv, nOK := newMap[k]
			switch {
			case !oOK:
				*out = append(*out, Difference{"response", strings.Join(child, "."), absent, NormValue(nv)})
			case !nOK:
				*out = append(*out, Difference{"response", strings.Join(child, "."), NormValue(ov), absent})
			default:
				diffJSON(ov, nv, child, out)
			}
		}
		return
	}
	oldArr, oldIsArr := old.([]any)
	newArr, newIsArr := new.([]any)
	if oldIsArr && newIsArr {
		if len(oldArr) != len(newArr) {
			*out = append(*out, Difference{"response", joined() + ".length", len(oldArr), len(newArr)})
		}
		n := min(len(oldArr), len(newArr))
		for i := 0; i < n; i++ {
			diffJSON(oldArr[i], newArr[i], append(append([]string{}, path...), strconv.Itoa(i)), out)
		}
		return
	}
	if !equal(old, new) {
		*out = append(*out, Difference{"response", joined(), NormValue(old), NormValue(new)})
	}
}

// Options controls response comparison.
type Options struct {
	NoisePaths     []string
	CompareHeaders []string
	SortArrays     []SortRule
}

func DiffResponses(
	oldStatus int, oldBody string,
	newStatus int, newBody string,
	oldHeaders, newHeaders map[string]string,
	opts Options,
) []Difference {
	var diffs []Difference
	if oldStatus != newStatus {
		diffs = append(diffs, Difference{"response", "status", oldStatus, newStatus})
	}
	for _, h := range opts.CompareHeaders {
		ov, nv := oldHeaders[h], newHeaders[h]
		if ov != nv {
			diffs = append(diffs, Difference{"response", "header." + h, ov, nv})
		}
	}
	oldParsed, oldIsJSON := tryJSON(oldBody)
	newParsed, newIsJSON := tryJSON(newBody)
	if oldIsJSON || newIsJSON {
		normalize := func(v any) any {
			return SortArrays(StripNoise(v, opts.NoisePaths), opts.SortArrays)
		}
		diffJSON(normalize(oldParsed), normalize(newParsed), nil, &diffs)
	} else if oldBody != newBody {
		diffs = append(diffs, Difference{"response", "body", oldBody, newBody})
	}
	return diffs
}

// ---------------------------------------------------------------------------
// write-set diff
// ---------------------------------------------------------------------------

// WriteEntry is the canonical representation of one row change.
type WriteEntry struct {
	Table   string         `json:"table"`
	Op      string         `json:"op"`
	PK      any            `json:"pk"`
	Values  map[string]any `json:"values,omitempty"`  // INSERT / DELETE
	Changed map[string]any `json:"changed,omitempty"` // UPDATE: col -> {before, after}
}

func colIsNoise(table, column string, patterns []string) bool {
	bare := table
	if i := strings.LastIndex(table, "."); i >= 0 {
		bare = table[i+1:]
	}
	for _, pat := range patterns {
		pt, pc, found := strings.Cut(pat, ".")
		if !found {
			continue
		}
		if (pt == "*" || pt == bare || pt == table) && (pc == "*" || pc == column) {
			return true
		}
	}
	return false
}

// NormalizeWriteset produces a canonical, order-insensitive write-set:
// changed non-noise columns only for UPDATEs, rows keyed by their `id`.
func NormalizeWriteset(changes []writeset.RowChange, noiseColumns, ignoreTables []string) []WriteEntry {
	ignored := map[string]bool{}
	for _, t := range ignoreTables {
		ignored[t] = true
	}

	var out []WriteEntry
	for _, ch := range changes {
		bare := ch.Table
		if i := strings.LastIndex(bare, "."); i >= 0 {
			bare = bare[i+1:]
		}
		if ignored[bare] || ignored[ch.Table] {
			continue
		}

		clean := func(row map[string]any) map[string]any {
			if row == nil {
				return nil
			}
			m := map[string]any{}
			for k, v := range row {
				if !colIsNoise(ch.Table, k, noiseColumns) {
					m[k] = NormValue(v)
				}
			}
			return m
		}

		entry := WriteEntry{Table: bare, Op: ch.Op}
		pkSource := ch.After
		if ch.Op == "DELETE" {
			pkSource = ch.Before
		}
		if pkSource != nil {
			entry.PK = NormValue(pkSource["id"])
		}

		switch ch.Op {
		case "UPDATE":
			before, after := clean(ch.Before), clean(ch.After)
			changed := map[string]any{}
			for k := range union(before, after) {
				if !equal(before[k], after[k]) {
					changed[k] = map[string]any{"before": before[k], "after": after[k]}
				}
			}
			if len(changed) == 0 { // only noise columns changed
				continue
			}
			entry.Changed = changed
		case "INSERT":
			entry.Values = clean(ch.After)
		default: // DELETE
			entry.Values = clean(ch.Before)
		}
		out = append(out, entry)
	}

	sort.Slice(out, func(i, j int) bool { return canonical(out[i]) < canonical(out[j]) })
	return out
}

func union(a, b map[string]any) map[string]bool {
	keys := map[string]bool{}
	for k := range a {
		keys[k] = true
	}
	for k := range b {
		keys[k] = true
	}
	return keys
}

func (e WriteEntry) payload() map[string]any {
	if e.Changed != nil {
		return e.Changed
	}
	return e.Values
}

// DiffWritesets pairs entries by (table, op, pk) and reports divergences.
func DiffWritesets(oldWS, newWS []WriteEntry) []Difference {
	key := func(e WriteEntry) string {
		return e.Table + "|" + e.Op + "|" + canonical(e.PK)
	}
	label := func(e WriteEntry) string {
		return e.Table + "[" + e.Op + " pk=" + canonical(e.PK) + "]"
	}

	oldBy := map[string][]WriteEntry{}
	for _, e := range oldWS {
		oldBy[key(e)] = append(oldBy[key(e)], e)
	}
	newBy := map[string][]WriteEntry{}
	for _, e := range newWS {
		newBy[key(e)] = append(newBy[key(e)], e)
	}

	keys := map[string]bool{}
	for k := range oldBy {
		keys[k] = true
	}
	for k := range newBy {
		keys[k] = true
	}
	sorted := make([]string, 0, len(keys))
	for k := range keys {
		sorted = append(sorted, k)
	}
	sort.Strings(sorted)

	var diffs []Difference
	for _, k := range sorted {
		olds, news := oldBy[k], newBy[k]
		n := min(len(olds), len(news))
		for i := 0; i < n; i++ {
			o, nw := olds[i], news[i]
			if !equal(o.payload(), nw.payload()) {
				diffPayload(label(o), o.payload(), nw.payload(), &diffs)
			}
		}
		for _, extra := range olds[n:] {
			diffs = append(diffs, Difference{"writeset", label(extra), extra.payload(), absent})
		}
		for _, extra := range news[n:] {
			diffs = append(diffs, Difference{"writeset", label(extra), absent, extra.payload()})
		}
	}
	return diffs
}

func diffPayload(label string, old, new map[string]any, out *[]Difference) {
	for k := range union(old, new) {
		ov, oOK := old[k]
		nv, nOK := new[k]
		if !oOK {
			ov = absent
		}
		if !nOK {
			nv = absent
		}
		if !equal(ov, nv) {
			*out = append(*out, Difference{"writeset", label + "." + k, ov, nv})
		}
	}
	sort.Slice(*out, func(i, j int) bool { return (*out)[i].Path < (*out)[j].Path })
}

func VerdictOf(responseDiffs, writesetDiffs []Difference) string {
	switch {
	case len(responseDiffs) > 0 && len(writesetDiffs) > 0:
		return BothDiff
	case len(responseDiffs) > 0:
		return ResponseDiff
	case len(writesetDiffs) > 0:
		return WritesetDiff
	default:
		return Match
	}
}

File: poc/python/mockingbox/diff.py
"""Diff engine: HTTP response diff + write-set diff, both noise-aware.

Verdicts (per request):
    MATCH          responses and write-sets are equivalent
    RESPONSE_DIFF  response bodies/status differ
    WRITESET_DIFF  DB row changes differ
    BOTH_DIFF      both differ
    ERROR          one side failed to respond / harness error
"""

from __future__ import annotations

import base64
import datetime
import decimal
import json
from dataclasses import dataclass, field

from .binlog import RowChange

MATCH = "MATCH"
RESPONSE_DIFF = "RESPONSE_DIFF"
WRITESET_DIFF = "WRITESET_DIFF"
BOTH_DIFF = "BOTH_DIFF"
ERROR = "ERROR"


@dataclass
class Difference:
    kind: str  # "response" | "writeset"
    path: str
    old: object
    new: object

    def as_dict(self) -> dict:
        return {"kind": self.kind, "path": self.path, "old": self.old, "new": self.new}


@dataclass
class RequestResult:
    name: str
    request: str
    verdict: str
    differences: list[Difference] = field(default_factory=list)
    old_status: int | None = None
    new_status: int | None = None
    old_writes: int = 0
    new_writes: int = 0
    error: str | None = None
    old_writeset: list[dict] = field(default_factory=list)
    new_writeset: list[dict] = field(default_factory=list)


# --------------------------------------------------------------------------
# value normalization (binlog & JSON values -> comparable primitives)
# --------------------------------------------------------------------------


def norm_value(v: object) -> object:
    if isinstance(v, decimal.Decimal):
        # scale-insensitive: Decimal("50000") == Decimal("50000.0")
        return int(v) if v == v.to_integral_value() else float(v)
    if isinstance(v, (datetime.datetime, datetime.date, datetime.time)):
        return v.isoformat()
    if isinstance(v, datetime.timedelta):
        return str(v)
    if isinstance(v, (bytes, bytearray)):
        return base64.b64encode(bytes(v)).decode()
    if isinstance(v, dict):
        return {k: norm_value(x) for k, x in v.items()}
    if isinstance(v, (list, tuple)):
        return [norm_value(x) for x in v]
    return v


# --------------------------------------------------------------------------
# noise path matching for JSON bodies
#   pattern segments: literal | * (exactly one) | ** (zero or more)
# --------------------------------------------------------------------------


def path_matches(pattern: str, path: list[str]) -> bool:
    return _match(pattern.split("."), path)


def _match(pat: list[str], path: list[str]) -> bool:
    if not pat:
        return not path
    head, rest = pat[0], pat[1:]
    if head == "**":
        return any(_match(rest, path[i:]) for i in range(len(path) + 1))
    if not path:
        return False
    if head == "*" or head == path[0]:
        return _match(rest, path[1:])
    return False


def strip_noise(obj: object, patterns: list[str], _path: list[str] | None = None) -> object:
    """Return a copy of obj with noise-matching keys removed."""
    path = _path or []
    if isinstance(obj, dict):
        out = {}
        for k, v in obj.items():
            child = path + [str(k)]
            if any(path_matches(p, child) for p in patterns):
                continue
            out[k] = strip_noise(v, patterns, child)
        return out
    if isinstance(obj, list):
        return [strip_noise(v, patterns, path + [str(i)]) for i, v in enumerate(obj)]
    return obj


# --------------------------------------------------------------------------
# response diff
# --------------------------------------------------------------------------


def _try_json(body: str | None) -> object:
    if body is None:
        return None
    try:
        return json.loads(body)
    except (ValueError, TypeError):
        return body


def diff_json(old: object, new: object, path: list[str], out: list[Difference]) -> None:
    if isinstance(old, dict) and isinstance(new, dict):
        for k in sorted(set(old) | set(new)):
            child = path + [str(k)]
            if k not in old:
                out.append(Difference("response", ".".join(child), "<absent>", norm_value(new[k])))
            elif k not in new:
                out.append(Difference("response", ".".join(child), norm_value(old[k]), "<absent>"))
            else:
                diff_json(old[k], new[k], child, out)
        return
    if isinstance(old, list) and isinstance(new, list):
        if len(old) != len(new):
            out.append(
                Difference("response", ".".join(path) + ".length", len(old), len(new))
            )
        for i, (o, n) in enumerate(zip(old, new)):
            diff_json(o, n, path + [str(i)], out)
        return
    if norm_value(old) != norm_value(new):
        out.append(Difference("response", ".".join(path) or "$", norm_value(old), norm_value(new)))


def diff_responses(
    old_status: int,
    old_body: str | None,
    new_status: int,
    new_body: str | None,
    noise_paths: list[str],
    old_headers: dict[str, str] | None = None,
    new_headers: dict[str, str] | None = None,
    compare_headers: list[str] | None = None,
) -> list[Difference]:
    diffs: list[Difference] = []
    if old_status != new_status:
        diffs.append(Difference("response", "status", old_status, new_status))

    for header in compare_headers or []:
        ov = (old_headers or {}).get(header)
        nv = (new_headers or {}).get(header)
        if ov != nv:
            diffs.append(Difference("response", f"header.{header}", ov, nv))

    old_parsed, new_parsed = _try_json(old_body), _try_json(new_body)
    if isinstance(old_parsed, (dict, list)) or isinstance(new_parsed, (dict, list)):
        old_clean = strip_noise(old_parsed, noise_paths)
        new_clean = strip_noise(new_parsed, noise_paths)
        diff_json(old_clean, new_clean, [], diffs)
    elif (old_body or "") != (new_body or ""):
        diffs.append(Difference("response", "body", old_body, new_body))
    return diffs


# --------------------------------------------------------------------------
# write-set diff
# --------------------------------------------------------------------------


def _col_is_noise(table: str, column: str, patterns: list[str]) -> bool:
    bare_table = table.split(".")[-1]
    for pat in patterns:
        pt, _, pc = pat.partition(".")
        if (pt in ("*", bare_table, table)) and (pc in ("*", column)):
            return True
    return False


def normalize_writeset(
    changes: list[RowChange],
    noise_columns: list[str],
    ignore_tables: list[str],
) -> list[dict]:
    """Canonical, order-insensitive representation of a request's write-set.

    UPDATE keeps only genuinely-changed, non-noise columns (before -> after).
    Rows are keyed by their `id` column when present so old/new can be paired.
    """
    out: list[dict] = []
    for ch in changes:
        bare_table = ch.table.split(".")[-1]
        if bare_table in ignore_tables or ch.table in ignore_tables:
            continue

        def clean(row: dict | None, *, table: str = ch.table) -> dict | None:
            if row is None:
                return None
            return {
                k: norm_value(v)
                for k, v in row.items()
                if not _col_is_noise(table, k, noise_columns)
            }

        entry: dict = {"table": bare_table, "op": ch.op}
        pk_source = ch.after if ch.op != "DELETE" else ch.before
        entry["pk"] = norm_value((pk_source or {}).get("id"))

        if ch.op == "UPDATE":
            before, after = clean(ch.before) or {}, clean(ch.after) or {}
            changed = {
                k: {"before": before.get(k), "after": after.get(k)}
                for k in sorted(set(before) | set(after))
                if before.get(k) != after.get(k)
            }
            if not changed:  # only noise columns changed -> drop entirely
                continue
            entry["changed"] = changed
        elif ch.op == "INSERT":
            entry["values"] = clean(ch.after)
        else:  # DELETE
            entry["values"] = clean(ch.before)
        out.append(entry)

    out.sort(key=lambda e: json.dumps(e, sort_keys=True, default=str))
    return out


def diff_writesets(old_ws: list[dict], new_ws: list[dict]) -> list[Difference]:
    """Compare canonical write-sets. Pair by (table, op, pk) where possible."""
    diffs: list[Difference] = []

    def key(e: dict) -> tuple:
        return (e["table"], e["op"], json.dumps(e.get("pk"), default=str))

    old_by_key: dict[tuple, list[dict]] = {}
    for e in old_ws:
        old_by_key.setdefault(key(e), []).append(e)
    new_by_key: dict[tuple, list[dict]] = {}
    for e in new_ws:
        new_by_key.setdefault(key(e), []).append(e)

    for k in sorted(set(old_by_key) | set(new_by_key), key=str):
        olds, news = old_by_key.get(k, []), new_by_key.get(k, [])
        label = f"{k[0]}[{k[1]} pk={k[2]}]"
        for o, n in zip(olds, news):
            payload_o = o.get("changed") or o.get("values")
            payload_n = n.get("changed") or n.get("values")
            if payload_o != payload_n:
                _diff_payload(label, payload_o, payload_n, diffs)
        for extra in olds[len(news):]:
            diffs.append(
                Difference("writeset", label, extra.get("changed") or extra.get("values"), "<absent>")
            )
        for extra in news[len(olds):]:
            diffs.append(
                Difference("writeset", label, "<absent>", extra.get("changed") or extra.get("values"))
            )
    return diffs


def _diff_payload(label: str, old: dict | None, new: dict | None, out: list[Difference]) -> None:
    old, new = old or {}, new or {}
    for col in sorted(set(old) | set(new)):
        if old.get(col) != new.get(col):
            out.append(
                Difference("writeset", f"{label}.{col}", old.get(col, "<absent>"), new.get(col, "<absent>"))
            )


def verdict_of(response_diffs: list[Difference], writeset_diffs: list[Difference]) -> str:
    if response_diffs and writeset_diffs:
        return BOTH_DIFF
    if response_diffs:
        return RESPONSE_DIFF
    if writeset_diffs:
        return WRITESET_DIFF
    return MATCH
